### DISCUSSION TITLE
Issue parsing space in node binary path when using with Laravel Herd

### DISCUSSION THREAD

**Discussion by jdarkins** [spatie/browsershot#874] [Q&A] [answered]
I'm having some issues when trying to use Browershot in a setup running through Laravel Herd. I am trying to set the node & npm binary's when calling browsershot, and in my setup these are defined in the ENV file and then through a config file.

I'm passing in the Node / NPM binary paths from my config & env files like so:

```
Browsershot::html($view)
            ->setNodeBinary(config('browsershot.node_binary'))
            ->setNpmBinary(config('browsershot.npm_binary'))
```

My env values with the paths look like this:

```
BROWSERSHOT_NODE_BINARY="/Users/username/Library/Application Support/Herd/config/nvm/versions/node/v22.6.0/bin/node"
BROWSERSHOT_NPM_BINARY="/Users/username/Library/Application Support/Herd/config/nvm/versions/node/v22.6.0/bin/npm"
```

However with the above I get the following error: `No such file or directory sh: /Users/username/Library/Application: No such file or directory`

I have tried to add a backslash to escape the space character which has been suggested in other discussion about this issue, like so:

```
BROWSERSHOT_NODE_BINARY="/Users/username/Library/Application\ Support/Herd/config/nvm/versions/node/v22.6.0/bin/node"
BROWSERSHOT_NPM_BINARY="/Users/username/Library/Application\ Support/Herd/config/nvm/versions/node/v22.6.0/bin/npm"
```

But then my entire site drops and I get the following error:

```
The environment file is invalid!
Failed to parse dotenv file. Encountered an unexpected escape sequence at ["/Users/username/Library/Application\ Support/Herd/config/nvm/versions/node/v22.6.0/bin/node"].
```

My guess here is the .env file doesn't like parsing a backslash, so I've tried adding this path directly into the browsershot setup to see if it would work like so:

```
Browsershot::html($view)
            ->setNodeBinary("/Users/username/Library/Application\ Support/Herd/config/nvm/versions/node/v22.6.0/bin/node")
            ->setNpmBinary("/Users/username/Library/Application\ Support/Herd/config/nvm/versions/node/v22.6.0/bin/npm")
```

In this case I removed the lines from my env, which brings the site back up, but then I get a weird error when trying to generate a pdf with browsershot:

![Screenshot 2024-08-20 at 11 07 09](https://github.com/user-attachments/assets/63b9e067-40b6-4df0-ba8c-32c5c39a81a2)

There is a 200 response and no other information to go off of - I've tried generating multiple pdf's with fresh data etc so in theory it shouldn't be related to any sort of dodgy data on my end. But I'm very stumped!

Does anyone have any suggestions?

**Comment by jdarkins** [comment]
Okay so I have figured out that the latest 'Failed to load PDF document' is a known bug related to Puppeteer v23 changes (fixed here #870) - which means this should work if I pass the path directly into that method.

However, this isn't feasible in my project as every user has their own local path which is defined in the `.env` file and passed through configs. So my issue now is how can I parse the space in the string from the `.env`?

**Reply by antonsumilov** [reply, marked as answer]
Hi, in my case I got around this by creating a symlink in the project folder (or any directory without spaces). And just use that symlink path in your .env.
Or even you can create symlinks same as this package defaults (/usr/local/bin/node and /usr/local/bin/npm), if I'm not mistaken, and do not use set methods at all.

**Reply by jdarkins** [reply]
Yeah this is the way I got round it too, symlinked the herd binaries in the `/Users/username/Library/Application\ Support/Herd/config/nvm/versions/node/v22.6.0/bin/node` paths to the corresponding  `/usr/local/bin/node` path and all working
